src/polhemus_tf_broadcaster.cpp

```diff
@@ -0,0 +1,292 @@
+  liberty_send(handle, cmd);
+}
+
+int main(int argc, char** argv) {
+
+  int i, s, nstations_found, nstations_attached, nstations_broadcast;
+  double x_hs, y_hs, z_hs;
+  bool broadcast_stations[8];
+  struct usb_device *dev;
+  usb_dev_handle *handle;
+  buffer_t buf;
+  struct timeval tv;
+
+  // Setup polhemus
+  usb_init();
+
+  dev = find_device_by_id(VENDOR, PRODUCT);
+  if (!dev) {
+    fprintf(stderr, "Could not find the Polhemus Liberty device.\n");
+    abort();
+  }
+
+  handle = usb_open(dev);
+  if (!handle) {
+    fprintf(stderr, "Could not get a handle to the Polhemus Liberty device.\n");
+    abort();
+  }
+
+  if (!liberty_init(handle)) {
+    fprintf(stderr, "Could not initialize the Polhemus Liberty device.\n");
+    usb_close(handle);
+    return 1;
+  }
+
+  init_buffer(&buf);
+  liberty_send(handle, (char *)"f1\r"); // activate binary mode
+
+  nstations_found = request_num_of_stations(handle, &buf);
+  fprintf(stderr, "Found %d stations.\n\n", nstations_found);
+  if (nstations_found == 0) {
+    usb_close(handle);
+    return 1;
+  }
+
+  // Setup ros
+  ros::init(argc, argv, "polhemus_tf_broadcaster");
+  ros::NodeHandle nh("/polhemus_tf_broadcaster");
+
+  // Retrieve launch parameters
+  nstations_attached = nstations_broadcast = 0;
+  for (i = 0; i < 8; i++) {
+    bool station_attached, station_broadcast;
+    s = i + 1;
+
+    nh.getParam("/station_"+std::to_string(s)+"_attached", station_attached);
+    nh.getParam("/station_"+std::to_string(s)+"_broadcast", station_broadcast);
+
+    broadcast_stations[i] = station_broadcast;
+
+    if (station_attached) {
+      nstations_attached += 1;
+      fprintf(stderr, "  Station %d attached.\n", s);
+    }
+
+    if (station_broadcast) {
+      nstations_broadcast += 1;
+      fprintf(stderr, "  Broadcasting station %d.\n\n", s);
+    }
+    else if (station_attached) {
+      fprintf(stderr, "\n");
+    } 
+  }
+
+  if (nstations_found != nstations_attached) {
+    fprintf(stderr, "Number of stations found (%d) does not match the number of stations attached (%d).\n", nstations_found, nstations_attached);
+    usb_close(handle);
+    return 1;
+  }
+
+  /* define which information to get per sensor (called a station
+     by polhemus)
+
+     o* applies to all stations
+
+     if this is changed, the euler_station_t struct or (***) below has to be edited accordingly 
+  */
+  liberty_send(handle, (char *)"O*,8,9,11,3,5\r"); // euler_station_t: euler angles
+
+  /* set output hemisphere -- this will produce a response which we're
+     ignoring 
+  */
+  nh.getParam("/x_hs", x_hs);
+  nh.getParam("/y_hs", y_hs);
+  nh.getParam("/z_hs", z_hs);
+  set_hemisphere(handle, x_hs, y_hs, z_hs);
+
+  /* switch output to centimeters */
+  //liberty_send(handle, "u1\r");
+  liberty_clear_input(handle); //right now, we just ignore the answer
+
+  // (***)
+  euler_station_t *stations = (euler_station_t*) (malloc(sizeof(euler_station_t) * nstations_found));
```

Comment:
Use c++ ```new```/```delete``` instead of c ```malloc```/```free```.

Reply:
Even better, use a ```std::vector<euler_station_t>```

Reply:
@VladimirIvan Would using `std::vector<station_t>` require changes to `liberty.cpp` and `liberty.h`?

Reply:
No, you just create the vector and ``push_back`` any ``station_t`` handle you create. No need to pre-allocate the size of a STL container

Reply:
Actually, you were correct. Don't use a vector because it would require that change to liberty. Liberty appears to require a traditional C array.

Reply:
@VladimirIvan and @wxmerkt if I understand correctly, we have to keep `malloc`/`free` rather than c++ `new`/`delete` because the polhemus uses C arrays? If so, I think the driver is complete.